gh-108951: document how to terminate an `asyncio.TaskGroup` (#123837)
python/cpython: Doc/library/asyncio-task.rst

   Improved handling of simultaneous internal and external cancellations
   and correct preservation of cancellation counts.

> Group Cancellation

Reviewer: I wonder if we should just not use the word "cancel" at all in this section. Name the exception `StopTaskGroup`, use "stop" or "terminate" or "exit" in the docstrings, and rename this section "Terminating a Task Group".

Because what we do really isn't similar to cancellation -- it doesn't raise a `CancelledError` (except internally to cancel individual tasks) and as a consequence doesn't have the precise semantics of cancelation in asyncio.

Reviewer 2: I think that suggestion makes sense @gvanrossum. I agree with "Terminating a Task Group"

Reviewer 3: +1; or even "Aborting Task Group"

------------------

Cancelling an entire task group is not natively supported by the standard
library but can be achieved by adding a task to the group that raises an
exception and ignoring that exception:

.. code-block:: python

   import asyncio
   from asyncio import TaskGroup

   class CancelTaskGroup(Exception):
     """Exception raised to cancel a task group."""

   async def job(task_id, sleep_time):
      print(f'Task {task_id}: start')
      await asyncio.sleep(sleep_time)
      print(f'Task {task_id}: done')

   async def cancel_task_group():
      """A task that would cancel the group it belongs to."""

>       raise CancelTaskGroup()

Reviewer: I'd move this before 'job', since it's part of the infrastructure everyone needs, while 'job' and 'main' are just examples.

Also change 'would cancel' to 'cancels'.

And possibly rename to 'stop_task_group' since it doesn't use 'CancelledError' or '.cancel()'?

> Cancelling an entire task group is not natively supported by the standard
> library but can be achieved by adding a task to the group that raises an
> exception and ignoring that exception:
> 
> .. code-block:: python
> 
>    import asyncio
>    from asyncio import TaskGroup
> 
>    class CancelTaskGroup(Exception):
>        """Exception raised to cancel a task group."""
> 
>    async def stop_task_group():
>        """A task that cancels the group it belongs to."""
>        raise CancelTaskGroup()
> 
>    async def job(task_id, sleep_time):
>        print(f'Task {task_id}: start')
>        await asyncio.sleep(sleep_time)
>        print(f'Task {task_id}: done')
> 
>    async def main():
>        try:
>            async with TaskGroup() as group:
>                # ... spawn some tasks ...
>                group.create_task(job(1, 1))
>                group.create_task(job(2, 2))
>                # create a task that cancels the group after 1.5 seconds
>                await asyncio.sleep(1.5)
>                group.create_task(stop_task_group())
>        except* CancelTaskGroup:
>            pass

Reviewer 2: A suggestion for renaming.

```suggestion
Terminating a task group is not natively supported by the standard
library. Termination can be achieved by adding an exception-raising task to the task group and ignoring the raised exception:

.. code-block:: python

   import asyncio
   from asyncio import TaskGroup

   class TerminateTaskGroup(Exception):
       """Exception raised to terminate a task group."""

   async def force_terminate_task_group():
       """Used to force termination of a task group"""
       raise TerminateTaskGroup()

   async def job(task_id, sleep_time):
       print(f'Task {task_id}: start')
       await asyncio.sleep(sleep_time)
       print(f'Task {task_id}: done')

   async def main():
       try:
           async with TaskGroup() as group:
               # ... spawn some tasks ...
               group.create_task(job(1, 1))
               group.create_task(job(2, 2))
               # Sleep 1.5 seconds
               await asyncio.sleep(1.5)
               # Add an exception-raising task to force the task group to terminate
               group.create_task(force_terminate_task_group())
       except* TerminateTaskGroup:
           pass
```

>    asyncio.run(main())

Author: ```suggestion
   import asyncio
   from asyncio import TaskGroup

   class CancelTaskGroup(Exception):
       """Exception raised to cancel a task group."""

   async def stop_task_group():
       """A task that cancels the group it belongs to."""
       raise CancelTaskGroup()

   async def job(task_id, sleep_time):
       print(f'Task {task_id}: start')
       await asyncio.sleep(sleep_time)
       print(f'Task {task_id}: done')

   async def main():
       try:
           async with TaskGroup() as group:
               # ... spawn some tasks ...
               group.create_task(job(1, 1))
               group.create_task(job(2, 2))
               # create a task that cancels the group after 1.5 seconds
               await asyncio.sleep(1.5)
               group.create_task(stop_task_group())
       except* CancelTaskGroup:
           pass

   asyncio.run(main())
```

   Improved handling of simultaneous internal and external cancellations
   and correct preservation of cancellation counts.

Group Cancellation
------------------

Cancelling an entire task group is not natively supported by the standard
library but can be achieved by adding a task to the group that raises an

> exception and ignore that exception:

Reviewer: Maybe:
```suggestion
exception and ignoring that exception:
```

.. code-block:: python

   import contextlib
   from asyncio import TaskGroup

   class CancelTaskGroup(Exception):
      """Exception raised to cancel a task group."""

   @contextlib.asynccontextmanager
   async def CancellableTaskGroup():
      """A cancellable task group."""
      try:
          async with TaskGroup() as group:
              yield group
      except* CancelTaskGroup:

>           print('task group was cancelled')

Reviewer: I think this helper context manager is unnecessary -- it's an elaboration, but to convey the basic idea we don't need it. We just show an example like this:
```py
try:
    with Taskgroup() as tg:
        ...spawn tasks...
except* CancelTaskGroup:
    pass

   async def cancel_group(cancellable_task_group):
      """Create a task that will cancel the group it belongs to."""
      async def factory():
          raise CancelTaskGroup(cancellable_task_group)

>       await cancellable_task_group.create_task(factory())

Reviewer: Similarly, to make the example more compact, we could just show a minimal helper task:
```py
async def cancel_it(group):
    raise CancelTaskGroup()
```
and a snippet showing how to then cancel the group:
```py
tg.create_task(cancel_it())
```
The rest is an elaboration the user can come up with depending on their needs.

(Maybe also mention that if you're *in* a task of that group, you can just `raise CancelTaskGroup`.)

> The above helper functions can be used as follows:

Reviewer: This whole block looks way too long. I think all we need is the suggestion I made above.
